pkg/controller/sync/sync_controller.go

```diff
@@ -100,33 +135,132 @@ func (r *ReconcileSync) Reconcile(request reconcile.Request) (reconcile.Result,
 		return reconcile.Result{}, nil
 	}
 
+	reportMetrics := false
+	defer func() {
+		if reportMetrics {
+			if err := r.reporter.reportSyncDuration(time.Since(timeStart)); err != nil {
+				log.Error(err, "failed to report sync duration")
+			}
+
+			r.metricsCache.ReportSync(&r.reporter)
+
+			if err := r.reporter.reportLastSync(); err != nil {
+				log.Error(err, "failed to report last sync timestamp")
+			}
+		}
+	}()
+
 	instance := &unstructured.Unstructured{}
 	instance.SetGroupVersionKind(gvk)
-	if err := r.reader.Get(context.TODO(), unpackedRequest.NamespacedName, instance); err != nil {
+
+	err = r.reader.Get(context.TODO(), unpackedRequest.NamespacedName, instance)
+	if err != nil {
 		if errors.IsNotFound(err) {
 			// This is a deletion; remove the data
 			instance.SetNamespace(unpackedRequest.Namespace)
 			instance.SetName(unpackedRequest.Name)
+			syncKey := strings.Join([]string{instance.GetNamespace(), instance.GetName()}, "/")
 			if _, err := r.opa.RemoveData(context.Background(), instance); err != nil {
 				return reconcile.Result{}, err
 			}
+
+			r.metricsCache.deleteCache(syncKey)
+			reportMetrics = true
+
 			return reconcile.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
 
+	syncKey := strings.Join([]string{instance.GetNamespace(), instance.GetName()}, "/")
 	if !instance.GetDeletionTimestamp().IsZero() {
 		if _, err := r.opa.RemoveData(context.Background(), instance); err != nil {
 			return reconcile.Result{}, err
 		}
+
+		r.metricsCache.deleteCache(syncKey)
+		reportMetrics = true
+
 		return reconcile.Result{}, nil
 	}
 
 	r.log.V(logging.DebugLevel).Info("data will be added", "data", instance)
 	if _, err := r.opa.AddData(context.Background(), instance); err != nil {
+		r.metricsCache.addCache(syncKey, tags{
+			kind:   instance.GetKind(),
+			status: metrics.ErrorStatus,
+		})
+		reportMetrics = true
+
 		return reconcile.Result{}, err
 	}
 
+	r.metricsCache.addCache(syncKey, tags{
+		kind:   instance.GetKind(),
+		status: metrics.ActiveStatus,
+	})
+
+	r.metricsCache.addKind(instance.GetKind())
+
+	reportMetrics = true
+
 	return reconcile.Result{}, nil
 }
+
+func NewMetricsCache() *MetricsCache {
+	return &MetricsCache{
+		Cache:      make(map[string]tags),
+		KnownKinds: make(map[string]bool),
+	}
+}
+
+func (c *MetricsCache) addKind(key string) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	c.KnownKinds[key] = true
```

Comment:
thanks, added a comment

Reply:
We should probably add why this is okay to avoid alarming readers of the code.

Reply:
Did we resolve this? I still see `this is a known memory leak` without context as to why that is okay.

Reply:
i had `// need to know encountered kinds to reset metrics for that kind` as a comment before, anything else we should add?

Reply:
"Because the footprint is capped at the total number of kinds ever synced, we expect an upper bound of the total number of kinds in the cluster, which should be limited and slow-growing."

Or something like the above would let users know the behavior of the memory leak and what behaviors cause the footprint to grow.

Reply:
Maybe also mention that the footprint should naturally reset on Pod upgrade b/c the container restarts.

Reply:
added